File: app/src/main/java/amrabed/android/release/evaluation/MainActivity.java
package amrabed.android.release.evaluation;

import android.content.Intent;
import android.os.Bundle;
import android.os.PersistableBundle;
import android.widget.ImageView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import com.bumptech.glide.Glide;
import com.firebase.ui.auth.AuthUI;
import com.google.android.material.snackbar.Snackbar;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;

import java.util.Collections;
import java.util.List;

import amrabed.android.release.evaluation.edit.EditActivity;
import amrabed.android.release.evaluation.locale.LocaleManager;


/**
 * Main Activity
 */
public class MainActivity extends AppCompatActivity {
    public static final int SIGN_IN_REQUEST = 100;
    private static final int EDIT_REQUEST = 10;
    private FirebaseUser user;
    private NavigationDrawer drawer;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setTheme(R.style.AppTheme_NoActionBar);
        super.onCreate(savedInstanceState);
        LocaleManager.setLocale(this);
        setContentView(R.layout.main_activity);

        user = FirebaseAuth.getInstance().getCurrentUser();
        if (user == null) {
            startActivityForResult(createSignInIntent(), SIGN_IN_REQUEST);
        } else {
            showWelcomeMessage();
            updateProfilePicture();
        }

        final Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        drawer = new NavigationDrawer(this).create(savedInstanceState, toolbar);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState, @NonNull PersistableBundle outPersistentState) {
        super.onSaveInstanceState(outState, outPersistentState);
        drawer.saveState(outState);
    }

    @Override
    protected void onResume() {
        super.onResume();
        LocaleManager.setLocale(this);
    }

    @Override
    public void onBackPressed() {
        if (drawer.isOpen()) {
            drawer.close();
        } else {
            drawer.onBackStackChanged();
            super.onBackPressed();
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == SIGN_IN_REQUEST) {
            if (resultCode == RESULT_OK) { // Successfully signed in
                user = FirebaseAuth.getInstance().getCurrentUser();
                showWelcomeMessage();
                updateProfilePicture();
            } else {
                Toast.makeText(this, R.string.no_sign_in, Toast.LENGTH_SHORT).show();
            }
        } else if (requestCode == EDIT_REQUEST) {
            if (resultCode == RESULT_OK) {
                recreate();
            }
        }
    }

    private void showWelcomeMessage() {
        if (user != null) {
            final String name = user.getDisplayName();
            final String text = (name != null) ?
                    getString(R.string.welcome) + " " + name.split(" ")[0] :
                    getString(R.string.signed_in);
            if (getCurrentFocus() != null) {
                Snackbar.make(getCurrentFocus(), text, Snackbar.LENGTH_SHORT).show();
            }
        }
    }

    private void updateProfilePicture() {
        Glide.with(this).load(user.getPhotoUrl()).into((ImageView) findViewById(R.id.user));
    }

    public void startEditorActivity() {
        startActivityForResult(new Intent(this, EditActivity.class), EDIT_REQUEST);
    }

    private static Intent createSignInIntent() {
        final List<AuthUI.IdpConfig> providers = Collections.singletonList(
                new AuthUI.IdpConfig.GoogleBuilder().build());
//                new AuthUI.IdpConfig.AnonymousBuilder().build());

        return AuthUI.getInstance()
                .createSignInIntentBuilder()
                .setAvailableProviders(providers)
                .setLogo(R.drawable.logo)
                .setTheme(R.style.AppTheme_FullScreen)
                .build();
    }

    public void signOut() {
        if (FirebaseAuth.getInstance().getCurrentUser() == null) {
            Toast.makeText(this, R.string.no_sign_in, Toast.LENGTH_SHORT).show();
        } else {
            AuthUI.getInstance().signOut(MainActivity.this)
                    .addOnCompleteListener(this, task -> {
                        Toast.makeText(this, R.string.signed_out, Toast.LENGTH_SHORT).show();
                        Glide.with(MainActivity.this).clear((ImageView) findViewById(R.id.user));
                    });
        }
    }
}

File: app/src/main/java/amrabed/android/release/evaluation/MainActivity.kt
package amrabed.android.release.evaluation

import amrabed.android.release.evaluation.databinding.MainActivityBinding
import amrabed.android.release.evaluation.utilities.auth.Authenticator
import android.app.Activity
import android.app.AlertDialog
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ImageView
import androidx.navigation.findNavController
import androidx.navigation.ui.AppBarConfiguration
import androidx.navigation.ui.NavigationUI.setupActionBarWithNavController
import androidx.navigation.ui.setupWithNavController
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.google.android.material.snackbar.Snackbar

/**
 * Main Activity
 */
class MainActivity : BaseActivity(), View.OnClickListener {
    private lateinit var binding: MainActivityBinding

    private val navController by lazy {
        findNavController(R.id.fragment).apply {
            addOnDestinationChangedListener { _, destination, _ ->
                binding.toolbar.visibility =
                    if (destination.id == R.id.taskEditor) View.GONE else View.VISIBLE
                binding.navigation.visibility =
                    if (destination.id == R.id.taskEditor) View.GONE else View.VISIBLE
                binding.user.visibility =
                    if (destination.id != R.id.assessment) View.GONE else View.VISIBLE
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = MainActivityBinding.inflate(layoutInflater)
        setContentView(binding.root)

        if (Authenticator.user != null) {
            Glide.with(this).load(Authenticator.user?.photoUrl)
                .apply(RequestOptions.circleCropTransform())
                .placeholder(R.drawable.ic_user).into(findViewById(R.id.user))
        }

        val appBarConfiguration =
            AppBarConfiguration(setOf(R.id.assessment, R.id.progress, R.id.guide))
        setSupportActionBar(binding.toolbar)
        setupActionBarWithNavController(this, navController, appBarConfiguration)
        binding.toolbar.setupWithNavController(navController, appBarConfiguration)
        binding.navigation.setupWithNavController(navController)
    }

    override fun setTitle(title: CharSequence?) {
        binding.toolbar.title = title
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.main, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.listEditor -> {
                startActivityForResult(Intent(this, EditActivity::class.java), EDIT_REQUEST)
                return true
            }
            R.id.settings -> {
                startActivity(Intent(this, SettingsActivity::class.java))
                return true
            }
            R.id.help -> {
                startActivity(Intent(Intent.ACTION_VIEW).apply {
                    data = Uri.parse(getString(R.string.helpWebsite))
                })
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onClick(v: View?) {
        // Profile picture clicked -> sign out if signed in
        if (Authenticator.user != null) {
            AlertDialog.Builder(this).setMessage(R.string.confirmSignOut)
                .setNegativeButton(R.string.no, null)
                .setPositiveButton(R.string.yes) { _, _ ->
                    Authenticator.signOut(this) {
                        Glide.with(this).clear(findViewById<ImageView>(R.id.user))
                        Snackbar.make(
                            window.decorView.rootView,
                            R.string.signedOut,
                            Snackbar.LENGTH_SHORT
                        ).show()
                    }
                }
                .create().show()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == EDIT_REQUEST && resultCode == Activity.RESULT_OK) {
            recreate()
        }
    }

    companion object {
        private const val EDIT_REQUEST = 10
    }
}